src/clj/rems/api/applications_v2.clj

```diff
@@ -0,0 +1,869 @@
+    (is (= {:permissions/by-role {:unrelated #{:foo}
+                                  :role #{:gazonk}}}
+           (-> {}
+               (set-role-permissions {:unrelated [:foo]
+                                      :role [:bar]})
+               (set-role-permissions {:role [:gazonk]}))))))
+
+(defn- user-permissions
+  "Returns the specified user's permissions to this application.
+   Union of all role and user specific permissions. Read the source
+   to find out the supported roles. See also `set-permissions`."
+  [application user-id]
+  (let [applicant? (= user-id (:application/applicant application))
+        handler? (contains? (:workflow.dynamic/handlers application) user-id)
+        permissions (remove nil?
+                            [(get-in application [:permissions/by-user user-id])
+                             (when applicant?
+                               (get-in application [:permissions/by-role :applicant]))
+                             (when handler?
+                               (get-in application [:permissions/by-role :handler]))])]
+    (if (empty? permissions)
+      nil
+      (apply set/union permissions))))
+
+(deftest test-user-permissions
+  (testing "no access"
+    (is (= nil
+           (user-permissions {}
+                             "user"))))
+  (testing "read-only access"
+    (is (= #{}
+           (user-permissions {:permissions/by-user {"user" #{}}}
+                             "user"))))
+  (testing "user permissions"
+    (is (= #{:foo}
+           (user-permissions {:permissions/by-user {"user" #{:foo}}}
+                             "user"))))
+  (testing "applicant permissions"
+    (is (= #{:foo}
+           (user-permissions {:application/applicant "user"
+                              :permissions/by-role {:applicant #{:foo}}}
+                             "user"))))
+  (testing "handler permissions"
+    (is (= #{:foo}
+           (user-permissions {:workflow.dynamic/handlers #{"user"}
+                              :permissions/by-role {:handler #{:foo}}}
+                             "user"))))
+  (testing "combined permissions from multiple roles"
+    (let [application {:application/applicant "user"
+                       :workflow.dynamic/handlers #{"user"}
+                       :permissions/by-user {"user" #{:foo}}
+                       :permissions/by-role {:applicant #{:bar}
+                                             :handler #{:gazonk}}}]
+      (is (= #{:foo :bar :gazonk}
+             (user-permissions application "user")))
+      (is (= nil
+             (user-permissions application "wrong user"))))))
+
+;;;; v2 API, pure application state based on application events
+
+(defmulti ^:private application-view-specific
+  "See `application-view`"
+  (fn [_application event] (:event/type event)))
+
+(defmethod application-view-specific :application.event/created
+  [application event]
+  (-> application
+      (assoc :application/id (:application/id event)
```

Comment:
Why are all the keys top-level? Why not have a hierarchy?

Reply:
Because it's not needed when we use namespaced keys: https://github.com/CSCfi/rems/blob/master/docs/architecture/003-namespaced-keywords.md

Reply:
I think that disagrees with my idea of how Clojure data structures are used.

Reply:
What do you propose? Something like this?

```clj
{:application/id 1
 :application/created (DateTime. 1000)
 :application/modified (DateTime. 1000)
 :application/last-activity (DateTime. 1000)
 :application/applicant "applicant"
 :application/applicant-attributes {"eppn" "applicant"
                                    "mail" "applicant@example.com"
                                    "commonName" "Applicant"}
 :application/resources [{:catalogue-item/id 10
                          :resource/id 11
                          :resource/ext-id "urn:11"
                          :catalogue-item/title {:en "en title"
                                                 :fi "fi title"
                                                 :default "non-localized title"}
                          :catalogue-item/start (DateTime. 100)
                          :catalogue-item/state :enabled}
                         {:catalogue-item/id 20
                          :resource/id 21
                          :resource/ext-id "urn:21"
                          :catalogue-item/title {:en "en title"
                                                 :fi "fi title"
                                                 :default "non-localized title"}
                          :catalogue-item/start (DateTime. 100)
                          :catalogue-item/state :enabled}]
 :application/licenses [{:license/id 30
                         :license/accepted false
                         :license/type :link
                         :license/start (DateTime. 100)
                         :license/end nil
                         :license/title {:en "en title"
                                         :fi "fi title"
                                         :default "non-localized title"}
                         :license/link {:en "http://en-license-link"
                                        :fi "http://fi-license-link"
                                        :default "http://non-localized-license-link"}}
                        {:license/id 31
                         :license/accepted false
                         :license/type :text
                         :license/start (DateTime. 100)
                         :license/end nil
                         :license/title {:en "en title"
                                         :fi "fi title"
                                         :default "non-localized title"}
                         :license/text {:en "en license text"
                                        :fi "fi license text"
                                        :default "non-localized license text"}}]
 :application/events []
 :application/description ""
 :application/form {:form/id 40
                    :form/title "form title"
                    :form/fields [{:field/id 41
                                   :field/value ""
                                   :field/type :description
                                   :field/title {:en "en title" :fi "fi title"}
                                   :field/placeholder {:en "en placeholder" :fi "fi placeholder"}
                                   :field/optional false
                                   :field/options []
                                   :field/max-length 100}
                                  {:field/id 42
                                   :field/value ""
                                   :field/type :text
                                   :field/title {:en "en title" :fi "fi title"}
                                   :field/placeholder {:en "en placeholder" :fi "fi placeholder"}
                                   :field/optional false
                                   :field/options []
                                   :field/max-length 100}]}
 :application/workflow {:workflow/id 50
                        :workflow/type :dynamic
                        :workflow/state :rems.workflow.dynamic/draft
                        :workflow.dynamic/handlers #{"handler"}}
 :application/allowed-commands #{::dynamic/save-draft
                                 ::dynamic/submit}

```

Reply:
Form and workflow attributes are now nested.